### PR TITLE
Sophie

### PR DESCRIPTION
**Pull request by sop26**
shooter subsystem and command

### CODE REVIEW

File: src/main/cpp/subsystems/ControlBoard.cpp [hunk 1]
```diff
@@ -5,32 +5,84 @@
 #include "subsystems/ControlBoard.h"
 #include "CONSTANTS.h"
```
Comment (alisonsoong) [inline src/main/cpp/subsystems/ControlBoard.cpp:6]: move to header file

File: src/main/cpp/subsystems/ControlBoard.cpp [hunk 2]
```diff
@@ -5,27 +5,84 @@
...
```
Comment (alisonsoong) [inline src/main/cpp/subsystems/ControlBoard.cpp]: Move to header file

File: src/main/cpp/subsystems/ControlBoard.cpp [hunk 3]
```diff
@@ -5,27 +5,84 @@
...
+
+bool ControlBoard::GetFlywheelPrepDesired(){
+  return _flywheelPrepDesired;
+}
+
+
+void ControlBoard::Update(){
+  _flywheelDesired = IsDown(_flywheelButton, FLYWHEEL_BUTTON_PORT);
+	_flywheelPrepDesired = IsDown(_flywheelPrepButton, FLYWHEEL_PREP_BUTTON_PORT);
 }
```
Comment (alisonsoong) [inline src/main/cpp/subsystems/ControlBoard.cpp]: We're moving away from constantly reading/keeping track of the states of the buttons. Check this out: https://docs.wpilib.org/en/stable/docs/software/commandbased/binding-commands-to-triggers.html 

ControlBoard can keep all the buttons in one place. Then, to schedule commands, we'll probably have to pass a reference of a button into RobotContainer..? 🤷 Still haven't thought about restructuring this a whole lot yet.

Also woah @rwlee, the custom triggers are so cool.

File: src/main/cpp/commands/ShooterCommand.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+ShooterCommand::ShooterCommand(Shooter& shooter,
+                                     ControlBoard& humanControl)
+    : _shooter{shooter}, _humanControl{humanControl} {
+        AddRequirements(&shooter);
+    }
+
+void ShooterCommand::Execute(){
+    if(_humanControl.GetFlywheelDesired()){
+        _shooter.SetFlywheelOutput(.75);
```
Comment (alisonsoong) [inline src/main/cpp/commands/ShooterCommand.cpp:15]: Need to be able to adjust flywheel speed using the dial. In my code, it was just reading the value off a "joystick," which just represents the dial.

File: src/main/cpp/commands/ShooterCommand.cpp [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "commands/ShooterCommand.h"
+
+ShooterCommand::ShooterCommand(Shooter& shooter,
+                                     ControlBoard& humanControl)
+    : _shooter{shooter}, _humanControl{humanControl} {
+        AddRequirements(&shooter);
+    }
+
+void ShooterCommand::Execute(){
```
Comment (alisonsoong) [inline src/main/cpp/commands/ShooterCommand.cpp:13]: And yeah, we want to move away from default commands—just want button triggered commands

File: src/main/cpp/subsystems/ControlBoard.cpp [hunk 6]
```diff
@@ -5,27 +5,84 @@
...
+}
+
+bool ControlBoard::GetFlywheelPrepDesired(){
+  return _flywheelPrepDesired;
+}
+
+
+void ControlBoard::Update(){
+  _flywheelDesired = IsDown(_flywheelButton, FLYWHEEL_BUTTON_PORT);
+	_flywheelPrepDesired = IsDown(_flywheelPrepButton, FLYWHEEL_PREP_BUTTON_PORT);
```
Comment (alisonsoong) [inline src/main/cpp/subsystems/ControlBoard.cpp]: Just putting a note here that we decided we no longer need to check for the state of ALL buttons all the time, only specific buttons during certain commands with multiple steps (like prepping -> shooting lol)

File: src/main/cpp/subsystems/RobotContainer.cpp [hunk 7]
```diff
@@ -19,4 +19,9 @@ void RobotContainer::ConfigureButtonBindings() {
 // An example command will be run in autonomous
 
 // declaring = know that they exist
-// defining = set what the things (that are defined) are equal to
\ No newline at end of file
+// defining = set what the things (that are defined) are equal to
+
+void RobotContainer::UpdateControls(){
+  _humanControl.Update();
```
Comment (alisonsoong) [inline src/main/cpp/subsystems/RobotContainer.cpp]: so no longer need this

File: src/main/cpp/subsystems/Shooter.cpp [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+#include "subsystems/Shooter.h"
+
+
+Shooter::Shooter() {
+    
+}
+
+void Shooter::SetFlywheelOutput(double power){
+    _flywheelMotor.Set(power);
+}
```
Comment (alisonsoong) [inline src/main/cpp/subsystems/Shooter.cpp:14]: might want to clamp power to be between 0 and 1

File: src/main/include/CONSTANTS.h [hunk 9]
```diff
@@ -5,3 +5,4 @@
 #pragma once
 
 constexpr double LEFT_TRIGGER_BARRIER = 0.5;
+constexpr double RIGHT_TRIGGER_BARRIER = 0.5;
```
Comment (alisonsoong) [inline src/main/include/CONSTANTS.h:18]: wait what's this for?

Comment (alisonsoong) [reply src/main/include/CONSTANTS.h:18]: Oh got it. Any reason why this is in a different file?

File: src/main/include/subsystems/RobotContainer.h [hunk 10]
```diff
@@ -18,11 +21,16 @@
...
 
  private:
   // The robot's subsystems and commands are defined here...
   Drivetrain _drivetrain{};
+  Shooter _shooter{};
   ControlBoard _humanControl{};
   DrivetrainCommand _drivetrainCommand{_drivetrain, _humanControl};
+  ShooterCommand _shooterCommand{_shooter, _humanControl};
   void ConfigureButtonBindings();
+  bool _flywheelDesired;
```
Comment (alisonsoong) [inline src/main/include/subsystems/RobotContainer.h:32]: Not exactly sure what this is for? I don't think it's being used anywhere.

File: src/main/include/subsystems/Shooter.h [hunk 11]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+#include <frc2/command/SubsystemBase.h>
+#include <frc/motorcontrol/PWMVictorSPX.h>
+#include "PORTS.h"
+#include "ControlBoard.h"
```
Comment (alisonsoong) [inline src/main/include/subsystems/Shooter.h:9]: I don't think ControlBoard is needed here?

File: src/main/cpp/subsystems/ControlBoard.cpp (hunk 3 above)
Comment (rwlee) [reply src/main/cpp/subsystems/ControlBoard.cpp]: Yeah, custom triggers are awesome -- so are multi button commands. If the operator uses an XBOX controller there's a lot of really cool opportunities. Operator control will be harder to train, but I think that's an acceptable sacrifice.

File: src/main/cpp/subsystems/ControlBoard.cpp (hunk 2 above)
Comment (rwlee) [reply src/main/cpp/subsystems/ControlBoard.cpp]: This one is actually ok, PORTS.h is used in the .h but CONSTANTS.h isn't needed in the header file (yet)

File: src/main/cpp/subsystems/ControlBoard.cpp (hunk 6 above)
Comment (rwlee) [reply src/main/cpp/subsystems/ControlBoard.cpp]: I really should talk about event handlers/listeners in better code.

File: src/main/include/CONSTANTS.h (hunk 9 above)
Comment (rwlee) [reply src/main/include/CONSTANTS.h:18]: Not sure... In general I'd like to have some organizational consistency -- we should decide how we want to organize things and put it into our style sheet.

My gut says:
ports -- rio ports, can ids, usb ids
robot constants -- mechanical constants like the number of modules, the physical location of camera, etc.
control constants -- sysid constants, PID constants, sensor values
human control constants -- deadband tuning, trigger barriers, etc.

I'm probably missing some there. Alternatively, subsystem specific constant files -- though that may come with it's own pitfalls

**Pull request closed without merging**